eras/shelley-ma/test-suite/test/Test/Cardano/Ledger/Mary/Value.hs

```diff
@@ -333,6 +340,86 @@ compactRoundTrip = forAll gen $ \v ->
+s2 = makeMultiAsset $ makeScriptHash "02"
+s3 = makeMultiAsset $ makeScriptHash "03"
+s4 = makeMultiAsset $ makeScriptHash "04"
+s5 = makeMultiAsset $ makeScriptHash "05"
+s6 = makeMultiAsset $ makeScriptHash "06"
+s7 = makeMultiAsset $ makeScriptHash "07"
+s8 = makeMultiAsset $ makeScriptHash "08"
+s9 = makeMultiAsset $ makeScriptHash "09"
+s10 = makeMultiAsset $ makeScriptHash "10"
+s11 = makeMultiAsset $ makeScriptHash "11"
+s12 = makeMultiAsset $ makeScriptHash "12"
+
+exampleMultiAssets :: [MultiAsset StandardCrypto]
+exampleMultiAssets = [s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12]
+
+-- | Test that the subtraction of Multi-assets (and the underlying 'CanonicalMaps')
+-- is a total function.
+-- This was used to diagnose https://github.com/input-output-hk/cardano-node/issues/4826
+subtractionIsTotal :: TestTree
+subtractionIsTotal = testProperty "marymary" $
+  QC.withMaxSuccess 100000 $
+    do
+      shuffle1 <- take 12 <$> QC.shuffle exampleMultiAssets
+      shuffle2 <- take 2 <$> QC.shuffle exampleMultiAssets
+      let a = mconcat [m | MultiAsset m <- shuffle1]
+          -- \^ here we chose to perform addition on the CanonicalMaps, as this is what
+          -- happens during deserialization, giving us insight into how node-4826 could
+          -- have occurred on mainnet (since we care about how the addition is associated).
+          -- Note that the ledger does not manipulate instances of
+          -- 'Value' and then store them in memory, since outputs are created by the user
+          -- and only deserialized. In other words, it is only in the ledger rules themselves
+          -- that we manipulate 'Value'.
+          b = mconcat shuffle2
+      pure $! rnf (MultiAsset a <> G.invert b)
+
+-- | The test below was discovered by a failure of 'subtractionIsTotal'
+-- using git sha bd359d3f745ca72242b2cd1208780c2787992b5f and --quickcheck-replay=649941
+node4826Reproducible :: TestTree
+node4826Reproducible =
+  testProperty "node4826Reproducible" $
+    let shuffle1 =
+          [ makeMultiAsset $ makeScriptHash "10"
+          , makeMultiAsset $ makeScriptHash "09"
+          , makeMultiAsset $ makeScriptHash "11"
+          , makeMultiAsset $ makeScriptHash "08"
+          , makeMultiAsset $ makeScriptHash "01"
+          , makeMultiAsset $ makeScriptHash "06"
+          , makeMultiAsset $ makeScriptHash "03"
+          , makeMultiAsset $ makeScriptHash "05"
+          , makeMultiAsset $ makeScriptHash "04"
+          , makeMultiAsset $ makeScriptHash "07"
+          , makeMultiAsset $ makeScriptHash "02"
+          , makeMultiAsset $ makeScriptHash "00"
+          ]
+        shuffle2 =
+          [ makeMultiAsset $ makeScriptHash "04"
+          , makeMultiAsset $ makeScriptHash "08"
+          ]
+        MultiAsset am = mconcat shuffle1
+        mm = mconcat [m | MultiAsset m <- shuffle1]
+        b = trace (unlines ["am ", show (valid am), showTree am, "mm:", show (valid mm), showTree mm]) $ mconcat shuffle2
```

Comment:
We should not leave any traces that will pollute the tests output. If this is something that needs to be printed upon a test failure then we can do that, but it cannot be done with tracing that is resorted purely for debugging. I don't really think it is worth it though, since we know the problem is fixed. So I suggest we remove this:
```suggestion
        b = mconcat shuffle2
```

Reply:
yea, I swear I fixed all that along with the boom == boom